pkg/reconciler/brokercell/broker_targets_config.go

```diff
@@ -0,0 +1,153 @@
+	brokerTargets := memory.NewEmptyTargets()
+	for _, broker := range brokers {
+		// Filter by `eventing.knative.dev/broker: <name>` here
+		// to get only the triggers for this broker. The trigger webhook will
+		// ensure that triggers are always labeled with their broker name.
+		triggers, err := r.triggerLister.Triggers(broker.Namespace).List(labels.SelectorFromSet(map[string]string{eventing.BrokerLabelKey: broker.Name}))
+		if err != nil {
+			logging.FromContext(ctx).Error("Failed to list triggers", zap.String("Broker", broker.Name), zap.Error(err))
+			bc.Status.MarkTargetsConfigFailed(configFailed, "failed to list triggers for broker %v: %v", broker.Name, err)
+			return err
+		}
+		r.addToConfig(ctx, broker, triggers, brokerTargets)
+	}
+	if err := r.updateTargetsConfig(ctx, bc, brokerTargets); err != nil {
+		logging.FromContext(ctx).Error("Failed to update broker targets configmap", zap.Error(err))
+		bc.Status.MarkTargetsConfigFailed(configFailed, "failed to update configmap: %v", err)
+		return err
+	}
+	bc.Status.MarkTargetsConfigReady()
+	return nil
+}
+
+// addToConfig reconstructs the data entry for the given broker and add it to targets-config.
+func (r *Reconciler) addToConfig(ctx context.Context, b *brokerv1beta1.Broker, triggers []*brokerv1beta1.Trigger, brokerTargets config.Targets) {
+	// TODO Maybe get rid of BrokerMutation and add Delete() and Upsert(broker) methods to TargetsConfig. Now we always
+	//  delete or update the entire broker entry and we don't need partial updates per trigger.
+	// The code can be simplified to r.targetsConfig.Upsert(brokerConfigEntry)
+	brokerTargets.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) {
+		// First delete the broker entry.
+		m.Delete()
+
+		// Then reconstruct the broker entry and insert it
+		m.SetID(string(b.UID))
+		m.SetAddress(b.Status.Address.URL.String())
+		m.SetDecoupleQueue(&config.Queue{
+			Topic:        brokerresources.GenerateDecouplingTopicName(b),
+			Subscription: brokerresources.GenerateDecouplingSubscriptionName(b),
+		})
+		if b.Status.IsReady() {
+			m.SetState(config.State_READY)
+		} else {
+			m.SetState(config.State_UNKNOWN)
+		}
+
+		// Insert each Trigger to the config.
+		for _, t := range triggers {
+			if t.Spec.Broker == b.Name {
+				target := &config.Target{
+					Id:        string(t.UID),
+					Name:      t.Name,
+					Namespace: t.Namespace,
+					Broker:    b.Name,
+					Address:   t.Status.SubscriberURI.String(),
+					RetryQueue: &config.Queue{
+						Topic:        brokerresources.GenerateRetryTopicName(t),
+						Subscription: brokerresources.GenerateRetrySubscriptionName(t),
+					},
+				}
+				if t.Spec.Filter != nil && t.Spec.Filter.Attributes != nil {
+					target.FilterAttributes = t.Spec.Filter.Attributes
+				}
+				if t.Status.IsReady() {
+					target.State = config.State_READY
+				} else {
+					target.State = config.State_UNKNOWN
+				}
```

Comment:
We may need to decouple the trigger state and the target state in the config. Our config should only reflect data plane readiness while trigger state means more I think.

Reply:
Added a TODO here. I'd like to keep the scope of this PR just a migration without adding more logic.